### PR TITLE
bpo-35922: Fix RobotFileParser when robots.txt is invalid

### PR DESCRIPTION
**Pull request by remilapeyre**
I don't think this change requires a news entry.

https://bugs.python.org/issue35922

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

File: Lib/test/test_robotparser.py [hunk 1]
```diff
@@ -96,6 +96,18 @@ class RejectAllRobotsTest(BaseRobotTest, unittest.TestCase):
     bad = ['/cyberworld/map/index.html', '/', '/tmp/']
 
 
+class InvalidFileTest(BaseRobotTest, unittest.TestCase):
```
Comment (taleinat) [inline Lib/test/test_robotparser.py]: This tests an empty file, which doesn't appear to be invalid from my reading. I suggest renaming to "EmptyFileTest".
```suggestion
class EmptyFileTest(BaseRobotTest, unittest.TestCase):
```

File: Lib/urllib/robotparser.py [hunk 2]
```diff
@@ -186,15 +186,17 @@ def crawl_delay(self, useragent):
         for entry in self.entries:
             if entry.applies_to(useragent):
                 return entry.delay
-        return self.default_entry.delay
+        if self.default_entry:
+            return self.default_entry.delay
```
Comment (taleinat) [inline Lib/urllib/robotparser.py:192]: Both `crawl_delay` and `request_rate` should return `None` explicitly if there is no applicable entry.

**Review by taleinat — CHANGES_REQUESTED** [review]
In addition to testing with empty robots.txt content, it would be good to also test with:
1. one not referencing the given user-agent and without a default
2. one with a default or with the given user-agent, but which doesn't define `Crawl_delay` and `Request_rate`

**Comment by remilapeyre** [comment]
Thanks for the review @taleinat! I made all the changes you requested, let me know if you want me to make other updates.

I have made the requested changes; please review again

**Review by taleinat — CHANGES_REQUESTED** [review]
These changes are very well done, and they were done very quickly; nice!

Taking a deeper look at the structure of the existing tests, we would do better to build on the existing `BaseRequestRateTest`. Specifically, I think its `test_request_rate()` method should be changed to always check the crawl delay and request rate, so that setting them to `None` will check that `RobotFileParser` actually returns `None`.

Doing so should allow removing `NoDefaultUserAgentTest`, since that case will already be covered by the existing `CrawlDelayAndRequestRateTest`. `EmptyFileTest` should stay but not require any special logic, just define `robots_txt` to an empty stings. Similarly, `NoRequestRateAndCrawlDelayTest` could be written much more simply.

Let me know if you'd like to do this yourself or if you'd prefer me to make these changes myself.

**Comment by remilapeyre** [comment]
Yeah, it will be better that way, I will push a new commit tomorrow.

**Comment by remilapeyre** [comment]
Thanks @taleinat, I have made the requested changes; please review again

File: Lib/test/test_robotparser.py [hunk 3]
```diff
@@ -96,17 +96,31 @@ class RejectAllRobotsTest(BaseRobotTest, unittest.TestCase):
     bad = ['/cyberworld/map/index.html', '/', '/tmp/']
 
 
+class EmptyFileTest(BaseRobotTest, unittest.TestCase):
+    robots_txt = ""
+
+
+class NoRequestRateAndCrawlDelayTest(BaseRobotTest, unittest.TestCase):
```
Comment (taleinat) [inline Lib/test/test_robotparser.py]: I believe this case is already covered by other tests, e.g. `RejectAllRobotsTest`.

File: Lib/test/test_robotparser.py [hunk 4]
```diff
@@ -141,10 +155,8 @@ class CrawlDelayAndRequestRateTest(BaseRequestRateTest, unittest.TestCase):
 
 class DifferentAgentTest(CrawlDelayAndRequestRateTest):
     agent = 'FigTree Robot libwww-perl/5.04'
-    # these are not actually tested, but we still need to parse it
-    # in order to accommodate the input parameters
-    request_rate = None
-    crawl_delay = None
+    request_rate = urllib.robotparser.RequestRate(9, 30)
+    crawl_delay = 3
```
Comment (taleinat) [inline Lib/test/test_robotparser.py]: These can simply be removed, since these are the same values as the base class.

**Review by taleinat — CHANGES_REQUESTED** [review]

**Comment by remilapeyre** [comment]
Thanks @taleinat for the review, ISTM that I should add `    good = ['/foo']` to `EmptyFileTest` right?

**Comment by taleinat** [comment]
> ISTM that I should add ` good = ['/foo']` to `EmptyFileTest` right?

Yes, that would be good.

**Comment by remilapeyre** [comment]
Thanks @taleinat, I think it should be good now (How many mistakes can I make on a simple patch ;) ?)

I have made the requested changes; please review again

**Comment by taleinat** [comment]
> I think it should be good now

I think so too! I'll make a final review later.

> Thanks @taleinat, I think it should be good now (How many mistakes can I make on a simple patch ;) ?)

Heh, the sky is the limit... But really, these aren't mistakes, this is the natural process for making changes: Most changes need at least a few iterations before they can go in. Additionally, in this particular case, I was working with this part of the codebase for the first time, so I was also learning as we went.

**Comment by taleinat** [comment]
@remilapeyre, this does need a NEWS file, plus a few minor fixes. For some reason I can't push to this PR's branch, nor create a PR for this branch in your repo. Perhaps you've got some privacy settings set in a non-standard way?

If you'd rather leave things the way they are, I'll create a new PR and still make sure to give you credit for this work.

**Comment by remilapeyre** [comment]
> Perhaps you've got some privacy settings set in a non-standard way?
The "Allow edits from maintainers." box is checked and I don't think I changed anything. I granted you rights on my fork, let me know if you're still blocked.

**Comment by taleinat** [comment]
> 
> 
> > Perhaps you've got some privacy settings set in a non-standard way?
> 
> The "Allow edits from maintainers." box is checked and I don't think I changed anything. I granted you rights on my fork, let me know if you're still blocked.

Ah, sorry, the problem was on my side; fixed now!

**Review by taleinat — APPROVED** [review]

**Comment by remilapeyre** [comment]
Not sure what the status check is, @taleinat should I backport manually?

**Comment by taleinat** [comment]
> 
> 
> Not sure what the status check is, @taleinat should I backport manually?

No, please ignore those messages. They seem to be caused by a bug in Miss Islington.

**Comment by taleinat** [comment]
> 
> 
> > Not sure what the status check is, @taleinat should I backport manually?
> 
> No, please ignore those messages. They seem to be caused by a bug in Miss Islington.

Or maybe Miss Islington was just made aware of the failure earlier...

**Comment by ZackerySpytz** [comment]
The commit message should use "robots.txt", not "robot.txt".

File: Lib/urllib/robotparser.py [hunk 5]
```diff
@@ -186,15 +186,21 @@ def crawl_delay(self, useragent):
...
+        else:
```
Comment (berkerpeksag) [inline Lib/urllib/robotparser.py]: The `else` branch isn't needed:

```py
if self.default_entry:
    return self.default_entry.delay
return None
```

File: Lib/urllib/robotparser.py [hunk 6]
```diff
@@ -186,15 +186,21 @@ def crawl_delay(self, useragent):
...
     def request_rate(self, useragent):
         if not self.mtime():
             return None
         for entry in self.entries:
             if entry.applies_to(useragent):
                 return entry.req_rate
-        return self.default_entry.req_rate
+        if self.default_entry:
+            return self.default_entry.req_rate
+        else:
```
Comment (berkerpeksag) [inline Lib/urllib/robotparser.py]: Same as above.

File: Misc/NEWS.d/next/Library/2019-06-11-19-34-29.bpo-35922.rxpzWr.rst [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Fix :meth:`RobotFileParser.crawl_delay()` and
```
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2019-06-11-19-34-29.bpo-35922.rxpzWr.rst]: Nit: () can be removed. Sphinx will it add it automatically.

File: Misc/NEWS.d/next/Library/2019-06-11-19-34-29.bpo-35922.rxpzWr.rst [hunk 8]
```diff
@@ -0,0 +1,4 @@
...
+:meth:`RobotFileParser.request_rate()` to return ``None`` rather than
```
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2019-06-11-19-34-29.bpo-35922.rxpzWr.rst]: Nit: () can be removed. Sphinx will it add it automatically.

File: Lib/test/test_robotparser.py [hunk 9]
```diff
@@ -97,30 +97,38 @@ class RejectAllRobotsTest(BaseRobotTest, unittest.TestCase):
...
                     )
                     self.assertEqual(
-                        self.parser.request_rate(agent).seconds,
+                        parsed_request_rate.seconds,
                         self.request_rate.seconds
                     )
 
 
+class EmptyFileTest(BaseRequestRateTest, unittest.TestCase):
+    robots_txt = ""
```
Comment (berkerpeksag) [inline Lib/test/test_robotparser.py]: Nit: Please use single quotes.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]
Good catch! I just left very minor comments. Otherwise it looks good to me.

**Comment by taleinat** [comment]
I have made the requested changes; please review again.

**Comment by taleinat** [comment]
@berkerpeksag, just awaiting your final okay, and I'll go about backporting this.

**Review by berkerpeksag — APPROVED** [review]
Good catch! Thank you for fixing this.

**Pull request merged**